### PR TITLE
Improved the API to allow filtering applications by any allowed attributes with a single logical operation.

### PR DESCRIPTION
**Pull request by ravinduramesh**
## Purpose
Facilitate filtering applications by cleintID through application listing page.
Resolves [#10990](https://github.com/wso2-enterprise/asgardeo-product/issues/10990), [#11371](https://github.com/wso2-enterprise/asgardeo-product/issues/11371), [#13933](https://github.com/wso2/product-is/issues/13933)

## When should this PR be merged
Merge after merging [#4172](https://github.com/wso2/carbon-identity-framework/pull/4172)

After this PR is merged to the master branch, it's possible to develop the frontend to filter applications by clientID through the application listing page.

## Approach
The API is enhanced to facilitate filtering applications by any allowed attribute (mentioned in a HashMap) with using zero or one logical operation. The filter parameter is validated and separated into nodes to create a binary tree. Then the final where clause to perform the filtering is generated by concatenating values of the nodes in the validated binary tree.

### CODE REVIEW

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/ServerApplicationManagementService.java [hunk 1]
```diff
@@ -147,8 +151,8 @@ public class ServerApplicationManagementService {
 
     private static final Log log = LogFactory.getLog(ServerApplicationManagementService.class);
 
-    private static final List<String> SEARCH_SUPPORTED_FIELDS = new ArrayList<>();
-    private static final String APP_NAME = "name";
+    // Allowed filter attributes mapped to real field names
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/ServerApplicationManagementService.java:154]: End the comments with full stop

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/ServerApplicationManagementService.java [hunk 2]
```diff
@@ -174,41 +181,72 @@ public ApplicationListResponse getAllApplications(Integer limit, Integer offset,
...
+                        String formattedLeftNode = generateFilterStringForBackend(
+                                expressionLeftNode.getAttributeValue(),
+                                expressionLeftNode.getOperation(),
+                                expressionLeftNode.getValue());
+                        String formattedRightNode = generateFilterStringForBackend(
+                                expressionRightNode.getAttributeValue(),
+                                expressionRightNode.getOperation(),
+                                expressionRightNode.getValue());
+                        formattedFilter = formattedLeftNode + " " +
+                                operationNode.getOperation() + " " + formattedRightNode;
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/ServerApplicationManagementService.java:231]: What are the supported `operationNode.getOperation()` operations ? Do we support both AND / OR?
Don't we need to validate the value of the operation of the operational node

Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/java/org/wso2/carbon/identity/api/server/application/management/v1/core/ServerApplicationManagementService.java:231]: Also need to test out this filtering mechanism in every type of supported DBs (MySQL, h2, mssql, db2, postgres, oracle) in product-is

**Comment by kayathiri4** [comment]
We need to update the `applications.yaml` file as well. Refer https://github.com/wso2/identity-api-server/blob/master/components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.common/src/main/java/org/wso2/carbon/identity/api/server/application/management/common/ApplicationManagementConstants.java [hunk 3]
```diff
@@ -56,9 +56,10 @@ private ApplicationManagementConstants() {
     public enum ErrorMessage {
 
         // Client errors with codes defined at Application Management Layer.
```
Comment (kayathiri4) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.common/src/main/java/org/wso2/carbon/identity/api/server/application/management/common/ApplicationManagementConstants.java:58]: Shall we update the LLC license change in the modified files

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/gen/java/org/wso2/carbon/identity/api/server/application/management/v1/ApplicationListItem.java [hunk 4]
```diff
@@ -164,6 +165,22 @@ public void setAccessUrl(String accessUrl) {
...
+        this.inboundKey = inboundKey;
+        return this;
+    }
+
+    @ApiModelProperty(example = "sampleRandomString", value = "")
+    @JsonProperty("inboundKey")
+    @Valid
+    public String getInboundKey() {
+        return inboundKey;
+    }
```
Comment (kayathiri4) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/gen/java/org/wso2/carbon/identity/api/server/application/management/v1/ApplicationListItem.java:179]: Shall we add a new line after the method

File: components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml [hunk 5]
```diff
@@ -2248,6 +2248,9 @@ components:
         accessUrl:
           type: string
           example: 'https://example.com/app/login'
+        inboundKey:
+          type: string
+          example: 'sampleRandomString_123'
```
Comment (kayathiri4) [inline components/org.wso2.carbon.identity.api.server.application.management/org.wso2.carbon.identity.api.server.application.management.v1/src/main/resources/applications.yaml:2253]: shall we give a realtime example?

**Review by kayathiri4 — APPROVED** [review]

**Pull request merged**